### ISSUE TITLE
Webhook body empty

### ISSUE THREAD

**Issue by aitchjoe** [prometheus/alertmanager#2836]
**What did you do?**

We create a Openresty gateway between Alertmanager and Feishu (a Chinese IM service) for transform JSON posted from Alertmanager webhook receiver.

**What did you expect to see?**

When Openresty receive post request, the request body should not be empty.

**What did you see instead? Under which circumstances?**

Sometimes the request body is empty, we modified Openresty nginx config:

```
    log_format  custom '$remote_addr - $remote_user [$time_local] '
                       '"$request" $status  $body_bytes_sent "$http_referer" '
                       '"$http_user_agent" "$http_x_forwarded_for" "$request_body"';
    access_log    "/opt/bitnami/openresty/nginx/logs/access.log" custom;
```

So we are sure sometimes the request_body from alertmanager is empty, then we enable alertmanager debug log:

```
level=debug ts=2022-02-24T02:06:13.395Z caller=dispatch.go:473 component=dispatcher aggrGroup="{}:{namespace=\"debug\"}" msg=flushing alerts="[PodDisruptionBudgetAtLimit[783f55a][active] PodDisruptionBudgetAtLimit[b796782][active] PodDisruptionBudgetAtLimit[a307cb0][active]]"
level=debug ts=2022-02-24T02:06:13.815Z caller=dispatch.go:473 component=dispatcher aggrGroup="{}/{alertname=\"Watchdog\"}:{}" msg=flushing alerts=[Watchdog[e25963d][active]]
level=error ts=2022-02-24T02:06:16.458Z caller=dispatch.go:309 component=dispatcher msg="Notify for alerts failed" num_alerts=35 err="Default/webhook[0]: notify retry canceled due to unrecoverable error after 1 attempts: unexpected status code 400: http://openresty.example.com/receiver"
```

Our code (http://openresty.example.com/receiver) will return 400 if request body is empty, and we checked full log but all flusing alerts are not empty, and from log we cannot relate the error line to the exact flushing line, so we dont know under which circumstances (which alerts) this happened.

There is similar issue [The request body sent through webhook is empty](https://github.com/prometheus/alertmanager/issues/2664), but we have not tested email receiver. Because we cannot reopen that issue so created this one.

**Environment**

* System information:

```
sh-4.4$ uname -srm
Linux 4.18.0-305.19.1.el8_4.x86_64 x86_64
```

* Alertmanager version:

```
sh-4.4$ alertmanager --version
alertmanager, version 0.21.0 (branch: rhaos-4.8-rhel-8, revision: f2d089f0537f3f7f222186d8a3c68ea93aa47726)
  build user:       root@bbfc7e6e9399
  build date:       20211026-23:28:54
  go version:       go1.16.6
```

Because we use Alertmanager integrated in OpenShift, we cannot easily upgrade it.

* Prometheus version:

```
sh-4.4$ prometheus --version
prometheus, version 2.26.1 (branch: rhaos-4.8-rhel-8, revision: b1b4ae6235c1476db2fa796a8a2e58f0633544bd)
  build user:       root@9382299c39b4
  build date:       20211026-23:30:45
  go version:       go1.16.6
  platform:         linux/amd64
```

* Alertmanager configuration file:

```
global:
  resolve_timeout: 5m
  http_config: {}
  smtp_hello: localhost
  smtp_require_tls: true
  pagerduty_url: https://events.pagerduty.com/v2/enqueue
  opsgenie_api_url: https://api.opsgenie.com/
  wechat_api_url: https://qyapi.weixin.qq.com/cgi-bin/
  victorops_api_url: https://alert.victorops.com/integrations/generic/20131114/alert/
route:
  receiver: Default
  group_by:
  - namespace
  routes:
  - receiver: Watchdog
    match:
      alertname: Watchdog
  - receiver: Critical
    match:
      severity: critical
  group_wait: 30s
  group_interval: 5m
  repeat_interval: 12h
inhibit_rules:
- source_match:
    severity: critical
  target_match_re:
    severity: warning|info
  equal:
  - namespace
  - alertname
- source_match:
    severity: warning
  target_match_re:
    severity: info
  equal:
  - namespace
  - alertname
receivers:
- name: Critical
  webhook_configs:
  - send_resolved: true
    http_config: {}
    url: http://openresty.example.com/receiver
    max_alerts: 0
- name: Default
  webhook_configs:
  - send_resolved: true
    http_config: {}
    url: http://openresty.example.com/receiver
    max_alerts: 0
- name: Watchdog
  webhook_configs:
  - send_resolved: true
    http_config: {}
    url: http://openresty.example.com/receiver
    max_alerts: 0
templates: []
```

**Comment by LeonYanghaha** [comment]
Try another version. That's how I solve the problem. My architecture is very similar to yours. I also have openresty. my alertmanager version is 0.22.2。

**Comment by aitchjoe** [comment]
Thanks but because we use Alertmanager integrated in OpenShift, we cannot easily upgrade it. But at least we know it's most likely a bug and not our wrong config. So is there any maintainer to confirm this and close issue?

**Comment by simonpasquier** [comment]
It's difficult to understand what happens here but since @LeonYanghaha said it's fixed by 0.22.0 in the same configuration, I'm closing it.
